### DISCUSSION TITLE
Convert DOM to perlish object?

### DISCUSSION THREAD

**Discussion by hussam-qasem** [mojolicious/mojo#2101] [Q&A] [answered]
I have a server that returns a simple XML, e.g.:
```xml
<Device id="1234">
  <info name="modelid" value="omg"/>
  <info name="picture" value="http://192.168.1.1/images/lol.jpg"/>
  <info name="status" value="0"/>
</Device>
```

is there a Mojo way to convert this DOM to a perlish object? Ultimately, I'd like to render a JSON. Thank you!

**Comment by Grinnz** [comment, marked as answer]
Not automatically, XML does not map to a json-like data structure. Your best option is to use Mojo::DOM and knowledge of the specific format you are receiving to translate it into an object in the structure that makes sense to you.

**Reply by hussam-qasem** [reply]
Thank you @Grinnz,

Is it possible to return a hash from a Mojo::Collection?

I can do something like this:
```perl
my $hash;
for my $info ( $dom->find('info')->each ){
  $hash->{ $info->attr('name') } = $info->attr('value');
}
```

but it would be nice to do it the Mojo way, e.g. (but this doesn't work):
```perl
my $hash = $dom->find('info')->map( sub { $_->attr('name') => $_->attr('value') } );
```

**Reply by Grinnz** [reply]
You're close it should be just a matter of reinterpreting the list as a hash:

`my %hash = @{$dom->find('info')->map( sub { $_->attr('name') => $_->attr('value') } )};`
